src/libraries/System.Private.CoreLib/src/System/WeakReference.cs

```diff
@@ -44,7 +58,112 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
+                // without throwing an exception.
+                if (default(IntPtr) == h)
+                    return false;
+
+                bool result = GCHandle.InternalGet(h) != null;
+
+                // must keep the instance alive as long as we use the handle.
+                GC.KeepAlive(this);
+
+                return result;
+            }
+        }
+
+        //Gets the Object stored in the handle if it's accessible.
+        // Or sets it.
+        public virtual object? Target
+        {
+            get
+            {
+                IntPtr h = Handle;
+                // Should only happen for corner cases, like using a WeakReference from a finalizer.
+                // GC can finalize the instance if it becomes F-Reachable.
+                // That, however, cannot happen while we use the instance.
+                //
+                // A derived class will be finalized with an actual Finalize, but the finalizer queue is single threaded,
+                // thus the default implementation will never access Target concurrently with finalizing.
+                //
+                // There is a possibility that a derived type overrides the default finalizer and arranges concurrent access.
+                // There is nothing that we can do about that and a few other exotic ways to break this.
+                //
+                if (default(IntPtr) == h)
+                    return default;
+
+                object? target = GCHandle.InternalGet(h);
+
+                // must keep the instance alive as long as we use the handle.
+                GC.KeepAlive(this);
+
+                return target;
+            }
+
+            set
+            {
+                IntPtr h = Handle;
+                // Should only happen for corner cases, like using a WeakReference from a finalizer.
+                // See the comment in the getter.
+                if (default(IntPtr) == h)
+                    throw new InvalidOperationException(SR.InvalidOperation_HandleIsNotInitialized);
+
+                GCHandle.InternalSet(h, value);
+
+                // must keep the instance alive as long as we use the handle.
+                GC.KeepAlive(this);
+            }
+        }
+
+        // Free all system resources associated with this reference.
+        ~WeakReference()
+        {
+            // Note: While WeakReference is formally a finalizable type, the finalizer does not actually run.
+            //       Instead the instances are treated specially in GC when scanning for no longer strongly-reachable
+            //       finalizable objects.
+            //
+            // Unlike WeakReference<T> case, the instance could be of a derived type and
+            //       in such case it is finalized via a finalizer.
+
+// eager finalization is NYI on Mono
+#if !MONO
+            Debug.Assert(this.GetType() != typeof(WeakReference));
```

Comment:
Just to double-check... do the delivered types get the `HasEagerFinalizerFlag` flag? Is there any test covering it?

Reply:
I am not sure if there are tests. It is very rare to derive from WeakReference and arguably a questionable idea to allow that in the first place, but I guess it made more sense before generics. 

The assert only checks that nonderived WeakReference uses eager finalization and never comes here.

Reply:
Here are some cases:

https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/WeakHashtable.cs#L145

https://github.com/dotnet/runtime/blob/932e3bdea7a4025b53fcec57de06ba2c3bca9976/src/tests/GC/API/WeakReference/NullHandle.cs#L13

Reply:
Thanks for checking!